rust/lance-core/src/datatypes/field.rs

```diff
@@ -69,6 +80,135 @@ impl Field {
+            ));
+        }
+        if options.compare_metadata && self.metadata != expected.metadata {
+            differences.push(format!(
+                "metadata for `{}` did not match expected metadata",
+                self_name
+            ));
+        }
+        if self.children.len() != expected.children.len()
+            || !self
+                .children
+                .iter()
+                .zip(expected.children.iter())
+                .all(|(child, expected)| child.name == expected.name)
+        {
+            let self_children = self
+                .children
+                .iter()
+                .map(|child| child.name.clone())
+                .collect::<HashSet<_>>();
+            let expected_children = expected
+                .children
+                .iter()
+                .map(|child| child.name.clone())
+                .collect::<HashSet<_>>();
+            let missing = expected_children
+                .difference(&self_children)
+                .cloned()
+                .collect::<Vec<_>>();
+            let unexpected = self_children
+                .difference(&expected_children)
+                .cloned()
+                .collect::<Vec<_>>();
+            if missing.is_empty() && unexpected.is_empty() {
+                differences.push(format!(
+                    "`{}` field order mismatch, expected [{}] but was [{}]",
+                    self_name,
+                    expected
+                        .children
+                        .iter()
+                        .map(|child| child.name.clone())
+                        .collect::<Vec<_>>()
+                        .join(", "),
+                    self.children
+                        .iter()
+                        .map(|child| child.name.clone())
+                        .collect::<Vec<_>>()
+                        .join(", "),
+                ));
+            } else {
+                differences.push(format!(
+                    "`{}` had mismatched children, missing=[{}] unexpected=[{}]",
+                    self_name,
+                    missing.join(", "),
+                    unexpected.join(", ")
+                ));
+            }
+        } else {
+            differences.extend(self.children.iter().zip(expected.children.iter()).flat_map(
+                |(child, expected_child)| {
+                    child.explain_differences(expected_child, options, Some(&self_name))
+                },
+            ));
+        }
+        differences
+    }
+
+    pub fn explain_difference(
+        &self,
+        expected: &Field,
+        options: &SchemaCompareOptions,
+    ) -> Option<String> {
+        let differences = self.explain_differences(expected, options, None);
+        if differences.is_empty() {
+            None
+        } else {
+            Some(differences.join(", "))
+        }
+    }
+
+    pub fn compare_with_options(&self, other: &Field, options: &SchemaCompareOptions) -> bool {
```

Comment:
It looks like we don't compare field_ids, which is good in some cases and not in others. I think that should be an option.